kernel/safe_typing.ml

```diff
@@ -1186,15 +1186,16 @@ let end_module l restype senv =
   let () = check_current_label l mp in
   let () = check_empty_context senv in
   let mbids = List.rev_map fst params in
-  let mb = build_module_body params restype senv in
+  let mb, cst = build_module_body params restype senv in
   let newenv = Environ.set_opaque_tables oldsenv.env (Environ.opaque_tables senv.env) in
   let newenv = Environ.set_native_symbols newenv senv.env.Environ.native_symbols in
   let newenv = set_engagement_opt newenv senv.engagement in
+  (* XXX univ handling can maybe be simplified here *)
   let senv'=
     propagate_loads { senv with
       env = newenv;
-      univ = Univ.ContextSet.union senv.univ mb.mod_constraints} in
-  let newenv = Environ.push_context_set ~strict:true mb.mod_constraints senv'.env in
+      univ = Univ.ContextSet.add_constraints cst senv.univ} in
+  let newenv = Environ.push_context_set ~strict:true (Univ.LSet.empty,cst) senv'.env in
```

Comment:
This is the source of the problem. The `newenv` does not contain the constraints of the module at that point, because it is the environment before the module was opened. You need to reset its universe graph here in the same way that the opaque table and native symbols are set above.